### PR TITLE
How to build a simple router using window event listeners in JavaScript

### LINKED ISSUE #3307: Building a simple JavaScript router using Window Event Listeners
### Proposed title of article
Building a simple JavaScript router using Window Event Listeners

### Introduction paragraph (2-3 paragraphs):
In a JavaScript application, the router is a component that is responsible for routing requests to their appropriate view.

It is responsible for displaying a view, and for determining which view to display based on the request. Consider a case scenario in a single page application where the user is first logged in and then navigates to the home page.

The router works by watching for changes in the URL and displaying the appropriate document. Every time the URL changes, the router checks to see if the new URL matches any of the routes that have been defined. If it does, the router displays the appropriate view.

In this tutorial, we will create a simple JavaScript router to demonstrate the basic concepts of routing in a JavaScript application.

### Key takeaways: 
By the end of this tutorial, the reader will be able to:
1. Define a route in a JavaScript application.
2. Register templates and views.
3. Create a simple JavaScript router.
4. Use the router to display a view.

### References: 
N/A

### PR DESCRIPTION
**Pull request by nancy-maina**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [ ] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [ ] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes
This closes #3307

### CODE REVIEW

**Comment by geoffrey45** [comment]
@nancy-maina Please add a hero image for your article.

**Comment by nancy-maina** [comment]
@geoffrey45 I have added a hero image.

**Comment by ahmadmardeni1** [comment]
plagiarism check done ✅

**Comment by geoffrey45** [comment]
@nancy-maina Your article is 610 words minus the code. Please update your article to reach the minimum requirements. ie 750 words.

File: content/articles/js-router/index.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+```javascript
+window.addEventListener('load', router);
+window.addEventListener('hashchange', router);
+```
+
+At this point, the simple router is ready. Open the index.html file in your browser and you should see the home page. Now, click the about link and you should see the about page.
+
+### Conclusion
+In this tutorial, you created a simple router in JavaScript. Being a simple router it only handles basic route mapping. It lacks many features that are available in a more complex router. For example, it does not offer support for nested routes.
+
```
Comment (geoffrey45) [inline content/articles/js-router/index.md:156]: Please provide an example of such a router and leave a link to it.

File: content/articles/js-router/index.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+In your script.js file, create two objects to hold the routes and the template functions to be rendered.
+
+```javascript
+let routes = {};
+let templates = {};
+```
+
+Then create two functions to display the home page and the about page. You can use arrow functions to define these functions, but I'll use regular functions for simplicity.
+
```
Comment (geoffrey45) [inline content/articles/js-router/index.md:39]: Add a link to the arrow functions MDN docs page.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@nancy-maina Please check the comments below.

**Comment by geoffrey45** [comment]
@nancy-maina

**Comment by nancy-maina** [comment]
I hav made the changes @geoffrey45

File: content/articles/js-router/index.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+```javascript
+let app_div = document.getElementById('app');
+
+function home() {
+    let div = document.createElement('div');
+    let link = document.createElement('a');
+    link.href = '#about';
+    link.innerText = 'About';
```
Comment (geoffrey45) [inline content/articles/js-router/index.md:50]: @nancy-maina The link: `#about` throws an error. Please change it to `#/about`

File: content/articles/js-router/index.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    div.appendChild(link);
+
+    app_div.appendChild(div);
+};
+
+function about() {
+    let div = document.createElement('div');
+    let link = document.createElement('a');
+    link.href = '#home';
+    link.innerText = 'Home';
```
Comment (geoffrey45) [inline content/articles/js-router/index.md:62]: The link `#home` should be `#/`. There is no route defined as `home` in your code.

**Comment by nancy-maina** [comment]
Hello @geoffrey45

I have made all the changes that you requested. Please check now.

**Comment by WanjaMIKE** [comment]
Nice job guys. @nancy-maina @geoffrey45

**Comment by WanjaMIKE** [comment]
Merged #3835 into master

**Pull request merged**